### PR TITLE
Add API Docs reference support on hover

### LINKED ISSUE #36219: Show links to API docs when hovering over constructs
**Description:**
$subject

In VSCode when you write some HTML and hover over an HTML tag, it shows a link to `MDN Reference` for the hovered html tag. Similarly if we can provide links to BBEs or an official documentation on Ballerina keywords/constructs for users to visit and learn more, it will help a lot for developers to learn Ballerina quickly.

**Describe your problem(s)**
See description

**Describe your solution(s)**
See description

### PR DESCRIPTION
**Pull request by malinthar**
## Purpose
This effort focuses on adding a reference for relevant API docs when a user hovers over the following constructs.

1. records
2. classes
3. objecttypes
4. clients
5. listeners
6. functions
7. constants
8. annotations
9. errors
10. types
11. enums (Not supported yet)
12. variables

Fixes #36219 

## Approach
image

## Samples
sample

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [x] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaTextDocumentService.java [hunk 1]
```diff
@@ -172,7 +172,7 @@ public CompletableFuture<Hover> hover(HoverParams params) {
                 // Ignore the cancellation exception
             } catch (Throwable e) {
                 // Note: Not catching UserErrorException separately to avoid flooding error msgs popups
-                String msg = "Operation 'text/hover' failed!";
+                String msg = "Operation 'text/hover' failed!";
```
Comment (sanjana) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/BallerinaTextDocumentService.java:175]: Superfluous spaces at the end of the line.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/hover/APIDocReferenceBuilder.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+                               String moduleName,
+                               String version,
+                               HoverConstructKind constructType,
+                               String constructReference) {
+        StringBuilder url = new StringBuilder(BASE_URL)
+                .append(String.join("\\/", orgName, moduleName, version, constructType.value()));
+        if (MD_NAVIGATED_SET.contains(constructType)) {
+            return url + "#" + constructReference;
+        }
+        return url.append("\\/").append(constructReference).toString();
```
Comment (sanjana) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/hover/APIDocReferenceBuilder.java:43]: Better if a newline is added before the `return` statement.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/hover/HoverSymbolResolver.java [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    protected Optional<Symbol> transformSyntaxNode(Node node) {
+        return this.visit(node.parent());
+    }
+
+    private Optional<Symbol> visit(Node node) {
+        if (node == null || visitedNodes.contains(node)) {
+            return Optional.empty();
+        }
+        visitedNodes.add(node);
+        return node.apply(this);
```
Comment (sanjana) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/hover/HoverSymbolResolver.java:67]: Better if newlines are added.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/hover/HoverUtil.java [hunk 4]
```diff
@@ -75,12 +85,37 @@ public static Hover getHover(HoverContext context) {
...
+                return hoverObj;
+            }
+            String url = APIDocReferenceBuilder.build(moduleID.get().orgName(),
+                    moduleID.get().moduleName(), moduleID.get().version(), constructKind.get(),
+                    symbol.get().getName().get());
+            markupContent.setValue((content.isEmpty() ? "" : content + MarkupUtils.getHorizontalSeparator())
+                    + "[View API Docs](" + url + ")");
+            hoverObj.setContents(markupContent);
+        }
+        return hoverObj;
```
Comment (sanjana) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/hover/HoverUtil.java:118]: Better if a newline is added.

File: language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/hover/HoverUtil.java [hunk 5]
```diff
@@ -170,4 +205,21 @@ public static Hover getDescriptionOnlyHoverObject(Documentation documentation) {
         }
         return HoverUtil.getHoverObject(description);
```
Comment (sanjana) [inline language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/hover/HoverUtil.java:206]: Ditto.

Comment (malinthar) [reply language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/hover/HoverUtil.java:207]: Fixed

**Review by mohanvive — APPROVED** [review]

**Pull request merged**